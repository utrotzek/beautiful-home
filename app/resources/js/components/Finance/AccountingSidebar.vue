<template>
    <div class="accountingSidebar">
        <ImporterModal
            :show="importerVisible"
            :period="period"
            @close="toggleImporter"
            @loading="$emit('loading')"
            @loadingCompleted="$emit('loadingCompleted')"
            @dataImported="$emit('dataImported')"
        />

        <ConnectionModal
            :planning="createConnectionData.planningData"
            :accounting="createConnectionData.accountingData"
            :show="showModal"
            :desired-amount="createConnectionData.desiredAmount"
            @close="closeModal"
            @doAccountingPlanningConnection="doAccountingPlanningConnection"
        />
        <div class="row">
            <div class="col-8">
                <h2 class="d-none d-md-block">
                    Ums채tze
                </h2>
            </div>
            <div class="col-12 col-md-4 text-left text-md-right">
                <div
                    id="accountingButtons"
                    class="mb-3"
                >
                    <button
                        class="btn btn-outline-dark"
                        title="Neuen Umsatz-Eintrag erstellen"
                        @click="createNewAccounting"
                    >
                        <i class="fa fa-plus-circle"></i>
                        <span class="d-inline-block d-md-none">
                            Neuer Eintrag
                        </span>
                    </button>

                    <button
                        class="btn btn-outline-dark"
                        title="csv einlesen"
                        @click="toggleImporter"
                    >
                        <i class="fa fa-arrow-circle-up"></i>
                        <span class="d-inline-block d-md-none">
                            Datei hochladen
                        </span>
                    </button>
                </div>
            </div>
        </div>

        <div
            id="accountingSearch"
            class="mb-1"
        >
            <Search @searched="accountingSearched" />
        </div>
        <div class="filter mb-2 text-right">
            Erledigte:
            <toggle-button
                v-model="showFinishedAccountings"
                :width="toggleWidth"
                :color="{checked: '#919191', unchecked: '#E9ECEF'}"
            />
            Unerledigte:
            <toggle-button
                v-model="showUnfinishedAccountings"
                :width="toggleWidth"
                :color="{checked: '#919191', unchecked: '#E9ECEF'}"
            />
        </div>

        <div
            v-if="filteredAccounting.length > 0"
            class="accounting-elements"
        >
            <div
                v-for="(item) in filteredAccounting"
                :key="item.id"
            >
                <AccountingElement
                    :accounting-data="item"
                    :show-connect-target="accountingConnectable(item.id)"
                    :class="accountingClass(item.id)"
                    :year="period.year"
                    :month="period.month"
                    @updateData="updateAccounting"
                    @deleteAccounting="deleteAccounting"
                    @doConnection="doShowConnectionModal"
                    @deleteConnection="deleteConnection"
                    @updateConnectedPlanning="updateConnectedPlanning"
                />
            </div>
        </div>
        <div v-else-if="searchQuery.length > 0">
            <div
                class="alert alert-light"
                role="alert"
            >
                F체r den eingegebenen Suchbegriff wurden keine Eintr채ge gefunden.
            </div>
        </div>
        <div v-else>
            <div
                class="alert alert-info"
                role="alert"
            >
                Keine Ums채tze vorhanden
            </div>
        </div>
    </div>
</template>

<script>
import moment from "moment";
import Search from "../tools/Search";
import AccountingElement from "./AccountingElement";
import ConnectionModal from "./ConnectionModal";
import ImporterModal from "./Importer/ImporterModal";

export default {
    name: "AccountingSidebar",
    components: {
        Search,
        AccountingElement,
        ConnectionModal,
        ImporterModal
    },
    props: {
        period: {
            type: Object,
            required: true,
            default() {
                return null;
            }
        },
        accountingData: {
            type: Array,
            required: true,
            default() {
                return [];
            }
        },
        createConnectionData: {
            type: Object,
            required: true,
            default() {
                return {
                    enabled: false,
                    planningId: 0,
                    accountingId: 0,
                    planningTotalAmount: 0,
                    accountingTotalAmount: 0,
                    desiredAmount: 0,
                    accountingData: null,
                    planningData: null
                };
            }
        }
    },
    data() {
        return {
            searchQuery: "",
            showModal: false,
            importerVisible: false,
            temp: false,
            showFinishedAccountings: true,
            showUnfinishedAccountings: true,
            toggleWidth: 35
        };
    },
    computed: {
        sortedAccounting: function() {
            return _.orderBy(this.accountingData, ["isNew", "date"], ["desc", "desc"]);
        },
        filteredAccounting() {
            return this.sortedAccounting.filter(accounting => {
                let display = (accounting.title.toLowerCase().search(this.searchQuery.toLowerCase()) > -1);

                if (!display){
                    let formattedDate = moment(accounting.date).format("DD.MM.YYYY");
                    display = (formattedDate.search(this.searchQuery.toLowerCase()) > -1);
                }

                if (!this.showFinishedAccountings && accounting.remainingAmount === 0.0){
                    display = false;
                }

                if (!this.showUnfinishedAccountings && accounting.remainingAmount !== 0.0){
                    display = false;
                }

                return display;
            });
        }
    },
    methods: {
        toggleImporter(){
            this.importerVisible = !this.importerVisible;
        },
        createNewAccounting() {
            this.$emit("createNewAccounting");
        },
        accountingSearched(query) {
            this.searchQuery = query;
        },
        accountingClass(id) {
            if (this.createConnectionData.enabled && !this.accountingConnectable(id)) {
                return {"deactivated": true};
            } else {
                return {"activated": true};
            }
        },
        getAccountingById(id){
            let i=0;
            for(i=0; i < this.accountingData.length;i++){
                if (id === this.accountingData[i].id)  {
                    return this.accountingData[i];
                }
            }
        },
        accountingConnectable(id) {
            if (this.createConnectionData.enabled && this.createConnectionData.planningData !== null){
                let accountingAmount = this.getAccountingById(id).remainingAmount;
                let planningAmount = this.createConnectionData.planningData.totalAmount;

                if (accountingAmount !== 0.0){
                    if ((planningAmount < 0 && accountingAmount < 0) ||
                        (planningAmount > 0 && accountingAmount > 0)
                    ) {
                        return true;
                    }
                }
            }

            return false;
        },
        updateAccounting(updatedAccounting) {
            this.$emit("updateAccounting", updatedAccounting);
        },
        deleteAccounting(accountingToDelete){
            this.$emit("deleteAccounting", accountingToDelete);
        },
        deleteConnection(accountId, connectedElementToRemove){
            this.$emit("deleteConnection", accountId, connectedElementToRemove);
        },
        updateConnectedPlanning(updatedPlanningItem, accountingId) {
            this.$emit("updateConnectedPlanning", updatedPlanningItem, accountingId);
        },
        closeModal() {
            this.showModal = false;
        },
        doShowConnectionModal(accountId){
            let acounntingData = this.getAccountingById(accountId);
            let planningData = this.createConnectionData.planningData;

            this.createConnectionData.accountingData = acounntingData;
            this.createConnectionData.accountingId = accountId;
            this.createConnectionData.accountingTotalAmount = acounntingData.remainingAmount;
            this.createConnectionData.planningTotalAmount = planningData.totalAmount;

            //make sure the desired amount for the connection is not greater than the remaining amount
            //of the accounting element
            if (this.createConnectionData.accountingData.remainingAmount < 0){
                if (this.createConnectionData.accountingData.remainingAmount - planningData.totalAmount < 0) {
                    this.createConnectionData.desiredAmount = planningData.totalAmount;
                }else{
                    this.createConnectionData.desiredAmount = acounntingData.remainingAmount;
                }
            }

            if (this.createConnectionData.accountingData.remainingAmount > 0){
                if (this.createConnectionData.accountingData.remainingAmount - this.createConnectionData.planningTotalAmount > 0) {
                    this.createConnectionData.desiredAmount = this.createConnectionData.planningTotalAmount;
                }else{
                    this.createConnectionData.desiredAmount = this.createConnectionData.accountingData.remainingAmount;
                }
            }
            this.showModal = true;
        },
        doAccountingPlanningConnection(planning, accounting, desiredAmount){
            this.showModal = false;
            this.$emit("doAccountingPlanningConnection", planning, accounting, desiredAmount);
        },
    }
};
</script>

<style scoped>
    .filter label {
        padding: 0;
        margin: 0;
    }
</style>
