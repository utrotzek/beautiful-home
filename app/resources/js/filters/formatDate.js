import moment from "moment";

//TODO: set format in moment
export default function formatDate (value) {
    if (value) {
        if (typeof value === "object"){
            return moment(value.date).format("DD.MM.YYYY");
        }else{
            return moment(String(value)).format("DD.MM.YYYY");
        }
    }
}
