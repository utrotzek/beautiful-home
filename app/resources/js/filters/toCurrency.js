export default function toCurrency (value) {
    if (typeof value !== "number") {
        return value;
    }
    var formatter = new Intl.NumberFormat("de-DE", {
        style: "currency",
        currency: "EUR",
        minimumFractionDigits: 2
    });
    return formatter.format(value);
}
