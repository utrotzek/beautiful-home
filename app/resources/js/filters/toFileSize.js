export default function toFileSize(value){
    if (isNaN(value)){
        return "undefined size";
    }else{
        const units = [ "byte", "kB", "MB", "GB", "TB"];
        let newValue = value;
        let counter = 0;

        if (newValue >= 1000){
            while (newValue >= 1000){
                newValue = newValue / 1024;
                counter++;
            }
        }

        var formatter = new Intl.NumberFormat("de-DE", {
            minimumFractionDigits: 1,
            maximumFractionDigits: 1
        });
        return formatter.format(newValue) + " " + units[counter];
    }
}
