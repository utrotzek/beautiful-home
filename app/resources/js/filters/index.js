import toCurrency from "./toCurrency";
import formatDate from "./formatDate";
import toFileSize from "./toFileSize";

export default {
    install(Vue) {
        Vue.filter("toCurrency", toCurrency);
        Vue.filter("formatDate", formatDate);
        Vue.filter("toFileSize", toFileSize);
    }
};
