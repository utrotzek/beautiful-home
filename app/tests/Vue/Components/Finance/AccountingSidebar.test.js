import "../Mocks/MatchMedia.mock";
import moment from "moment";
import {shallowMount} from "@vue/test-utils";
import AccountingSidebar from "../../../../resources/js/components/Finance/AccountingSidebar";
import Vue from "vue";
import filters from "../../../../resources/js/filters";
import AccountingFixtures from "./Fixtures/AccountingFixtures";
import PlanningFixtures from "./Fixtures/PlanningFixtures";
import PeriodDataFixtures from "./Fixtures/PeriodFixtures";
import ToggleButton from "vue-js-toggle-button";
Vue.use(ToggleButton);
Vue.use(filters);

describe("Planning sidebar", () => {
    let wrapper = null;
    let accountingData = null;
    let planningData = null;
    let periodData = null;
    let costCenterFilter = [];
    let createConnectionData = {
        enabled: false,
        planningId: 0,
        accountingId: 0,
        planningTotalAmount: 0,
        accountingTotalAmount: 0,
        desiredAmount: 0,
        accountingData: null,
        planningData: null
    };

    beforeEach(() => {
        accountingData = AccountingFixtures;
        periodData = PeriodDataFixtures[0];
        planningData = PlanningFixtures;

        wrapper = shallowMount(AccountingSidebar,{
            propsData: {
                period: periodData,
                accountingData: accountingData,
                costCenterFilter: costCenterFilter,
                createConnectionData: createConnectionData
            }
        });

        wrapper.setData(
            {
                searchQuery: "",
                showModal: false,
                importerVisible: false,
                temp: false,
                showFinishedAccountings: true,
                showUnfinishedAccountings: true,
            }
        );
    });

    it ("renders properly", () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it ("Can create new accounting entries", async () => {
        wrapper.find(".new-accounting").trigger("click");
        await Vue.nextTick();
        expect(wrapper.emitted().createNewAccounting).toBeTruthy();
    });

    it ("Can toggle the importer", async () => {
        expect(wrapper.vm.importerVisible).toBeFalsy();
        wrapper.find(".toggle-importer").trigger("click");
        await Vue.nextTick();
        expect(wrapper.vm.importerVisible).toBeTruthy();
        wrapper.find(".toggle-importer").trigger("click");
        await Vue.nextTick();
        expect(wrapper.vm.importerVisible).toBeFalsy();
    });

    it ("entries can be searched by date", async () => {
        wrapper.vm.accountingSearched("02.01.3000");
        await Vue.nextTick();
        expect(wrapper.vm.filteredAccounting.length).toEqual(1);
        expect(wrapper.vm.filteredAccounting).toEqual([accountingData[1]]);
    });

    it ("date search can handle date objects", async () => {
        accountingData[1].date = moment("02.01.3000", "DD.MM.YYYY").toDate();
        wrapper.vm.accountingSearched("02.01.3000");
        await Vue.nextTick();
        expect(wrapper.vm.filteredAccounting.length).toEqual(1);
        expect(wrapper.vm.filteredAccounting).toEqual([accountingData[1]]);
    });

    it ("entries can be searched by title", async () => {
        wrapper.vm.accountingSearched("Rewe");
        await Vue.nextTick();
        expect(wrapper.vm.filteredAccounting.length).toEqual(1);
        expect(wrapper.vm.filteredAccounting).toEqual([accountingData[0]]);
    });

    it ("can apply cost center filter", async () => {
        accountingData[3].connectedPlanning[0].costCenter = null;
        accountingData[2].connectedPlanning = [];
        wrapper.setProps(
            {
                costCenterFilter: [
                    { id: 1, title: "Einkaufen"}
                ]
            }
        )
        await Vue.nextTick();
        expect(wrapper.vm.filteredAccounting.length).toEqual(2);
        expect(wrapper.vm.filteredAccounting).toContain(accountingData[1]);
        expect(wrapper.vm.filteredAccounting).toContain(accountingData[0]);
    });

    it ("can hide finished accountings", async () => {
        accountingData[0].remainingAmount = -100.0;
        accountingData[0].totalAmount = -100.0;
        wrapper.vm.showFinishedAccountings = false;
        expect(wrapper.vm.filteredAccounting.length).toEqual(1);
        expect(wrapper.vm.filteredAccounting).toContain(accountingData[0]);
    });

    it ("can hide unfinished accountings", async () => {
        accountingData[0].remainingAmount = -100.0;
        accountingData[0].totalAmount = -100.0;
        wrapper.vm.showUnfinishedAccountings = false;
        expect(wrapper.vm.filteredAccounting.length).toEqual(3);
        expect(wrapper.vm.filteredAccounting).not.toContain(accountingData[0]);
    });

    it ("connection mode will be disabled depending on remaining amount", async () => {
        createConnectionData.enabled = true;
        createConnectionData.planningData = {
            totalAmount: 0
        };
        expect(wrapper.vm.accountingClass(1)).toEqual({"deactivated": true});
    });

    it ("update accounting will fire event", async () => {
        wrapper.vm.updateAccounting(accountingData[0]);
        expect(wrapper.emitted().updateAccounting[0]).toEqual([accountingData[0]]);
    });

    it ("delete accounting will fire event", async () => {
        wrapper.vm.deleteAccounting(accountingData[0]);
        expect(wrapper.emitted().deleteAccounting[0]).toEqual([accountingData[0]]);
    });

    it ("delete connection will fire event", async () => {
        wrapper.vm.deleteConnection(1, {connection: "to remove"});
        expect(wrapper.emitted().deleteConnection[0]).toEqual([1, {connection: "to remove"}]);
    });

    it ("update connection will fire event", async () => {
        wrapper.vm.updateConnectedPlanning(planningData[0], 2);
        expect(wrapper.emitted().updateConnectedPlanning[0]).toEqual([planningData[0], 2]);
    });

    it ("can show connection modal", async () => {
        wrapper.vm.doShowConnectionModal(1);
        expect(wrapper.vm.showModal).toBeTruthy();
    });

    it ("can calculate desired amount properly when having negative amount", async () => {
        accountingData[0].remainingAmount = -100;
        createConnectionData = {
            enabled: true,
            accountingId: 1,
            planningData: planningData[0]
        };

        wrapper.setProps({createConnectionData: createConnectionData});
        wrapper.vm.doShowConnectionModal(1);
        expect(wrapper.vm.createConnectionData.desiredAmount).toEqual(-100);
    });

    it ("desired amount will be calculated properly when having positive amount", async () => {
        accountingData[0].remainingAmount = 100;
        planningData[0].totalAmount = 100;
        createConnectionData = {
            enabled: true,
            accountingId: 1,
            planningData: planningData[0]
        };

        wrapper.setProps({createConnectionData: createConnectionData});

        wrapper.vm.doShowConnectionModal(1);
        expect(wrapper.vm.createConnectionData.desiredAmount).toEqual(100);
    });

    it ("can hide connection modal", async () => {
        wrapper.vm.showModal = true;
        wrapper.vm.closeModal();
        expect(wrapper.vm.showModal).toBeFalsy();
    });

    it ("desired amount will match remaining amount even when having a higher negative planning amount", async () =>  {
        accountingData[0].remainingAmount = -200;
        planningData[0].totalAmount = -100;
        createConnectionData = {
            enabled: true,
            accountingId: 1,
            planningData: planningData[0]
        };

        wrapper.setProps({createConnectionData: createConnectionData});
        wrapper.vm.doShowConnectionModal(1);
        expect(wrapper.vm.createConnectionData.desiredAmount).toEqual(-100);
    });

    it ("desired amount will match remaining amount even when having a higher positive planning amount", async () =>  {
        accountingData[0].remainingAmount = 200;
        planningData[0].totalAmount = 100;
        createConnectionData = {
            enabled: true,
            accountingId: 1,
            planningData: planningData[0]
        };

        wrapper.setProps({createConnectionData: createConnectionData});
        wrapper.vm.doShowConnectionModal(1);
        expect(wrapper.vm.createConnectionData.desiredAmount).toEqual(100);
    });

    it ("can trigger accounting planning connection", async () => {
        const planning = planningData[0];
        const accounting = accountingData[0];
        const desiredAmount = -30;
        wrapper.vm.doAccountingPlanningConnection(planning, accounting, desiredAmount);

        expect(wrapper.emitted().doAccountingPlanningConnection[0]).toEqual([planning, accounting, desiredAmount]);
    });
});
