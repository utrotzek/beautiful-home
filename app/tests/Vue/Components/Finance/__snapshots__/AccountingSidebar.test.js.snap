// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Planning sidebar renders properly 1`] = `
<div
  class="accountingSidebar"
>
  <importermodal-stub
    period="[object Object]"
  />
   
  <connectionmodal-stub
    desiredamount="0"
  />
   
  <div
    class="row"
  >
    <div
      class="col-8"
    >
      <h2>
        
                Ums√§tze 
        <span
          class="fa fa-check-circle"
        />
      </h2>
    </div>
     
    <div
      class="col-12 col-md-4 text-left text-md-right"
    >
      <div
        class="mb-3"
        id="accountingButtons"
      >
        <button
          class="btn btn-outline-dark new-accounting"
          title="Neuen Umsatz-Eintrag erstellen"
        >
          <i
            class="fa fa-plus-circle"
          />
           
          <span
            class="d-inline-block d-md-none"
          >
            
                        Neuer Eintrag
                    
          </span>
        </button>
         
        <button
          class="btn btn-outline-dark toggle-importer"
          title="csv einlesen"
        >
          <i
            class="fa fa-arrow-circle-up"
          />
           
          <span
            class="d-inline-block d-md-none"
          >
            
                        Datei hochladen
                    
          </span>
        </button>
      </div>
    </div>
  </div>
   
  <div
    class="mb-1"
    id="accountingSearch"
  >
    <search-stub />
  </div>
   
  <div
    class="filter mb-2 text-right"
  >
    
        Erledigte:
        
    <toggle-button-stub
      color="[object Object]"
      height="16"
      margin="3"
      speed="300"
      value="true"
      width="30"
    />
    
        Unerledigte:
        
    <toggle-button-stub
      color="[object Object]"
      height="16"
      margin="3"
      speed="300"
      value="true"
      width="30"
    />
  </div>
   
  <div
    class="accounting-elements"
  >
    <div>
      <accountingelement-stub
        accountingdata="[object Object]"
        class="activated"
        month="1"
        year="3000"
      />
    </div>
    <div>
      <accountingelement-stub
        accountingdata="[object Object]"
        class="activated"
        month="1"
        year="3000"
      />
    </div>
    <div>
      <accountingelement-stub
        accountingdata="[object Object]"
        class="activated"
        month="1"
        year="3000"
      />
    </div>
    <div>
      <accountingelement-stub
        accountingdata="[object Object]"
        class="activated"
        month="1"
        year="3000"
      />
    </div>
  </div>
</div>
`;
